File: src/Result.ts
export class Result<T, E> {
  private readonly value: T | E;
  private readonly ok: boolean;

  private constructor(value: T | E, ok: boolean) {
    this.value = value;
    this.ok = ok;
  }

  public static ok<T, E>(value: T): Result<T, E> {
    return new Result<T, E>(value, true);
  }

  public static err<T, E>(error: E): Result<T, E> {
    return new Result<T, E>(error, false);
  }

  public map(ok: (value: T) => void, err: (error: E) => void): this {
    if (this.ok) {
      ok(this.value as T);
    } else {
      err(this.value as E);
    }

    return this;
  }
}

File: src/main/kotlin/io/frankmayer/dwf/lib/Result.kt
package io.frankmayer.dwf.lib

class Result<T, E> {
    private val value: T?
    private val error: E?
    private val isOk: Boolean

    companion object {
        fun <T, E> success(value: T): Result<T, E> {
            return Result(value, null, true)
        }

        fun <T, E> failure(error: E): Result<T, E> {
            return Result(null, error, false)
        }
    }

    private constructor(value: T?, error: E?, ok: Boolean) {
        this.value = value
        this.error = error
        this.isOk = ok
    }

    fun <R> map(okSuccess: (T) -> R, onFailure: (E) -> R): R {
        return if (isOk) {
            okSuccess(value!!)
        } else {
            onFailure(error!!)
        }
    }
}
